Add breadcrumbs to removal of kube objects

jansav opened the pull request:
Contains proposal to improve quality of codebase while implementing a trivial feature. It's a tool to make communication of difficult topics easier. Please read commits from start to finish as they describe a chain of thought.

Why?
- Because global shared state is bad.
- Because compliance to SOLID is a good way to improve design and quality overall.
- Because currently unit testing in codebase is too difficult and lacks infrastructure. Having unit tests enables faster development, fewer bugs, ability to refactor and confident deployments etc.

What was done?
- We implemented a trivial feature.
- In said feature, we replaced global shared state with dependencies.
- We implemented “local” unit tests which benefit from said dependencies that can now be stubbed.
- We implemented naive, staged, way of injecting a component with dependencies in place. This was done quick-to-dirty to better communicate a difficult topic with an example.
- We tidied up the dependencies to be minimal for the intended purpose, ie. compliance to Interface Segregation Principle.
- We introduced snapshot-testing for UI.
- We introduced library “asyncFn” to make notoriously difficult async unit testing easier.
- We introduced shared configuration for WebStorm users.

Guidelines:
- Solution must be simple and not scary.
- Solution must comply to well known patterns.
- Solution must be adaptable in small increments.

Talking points:
- Using test fakes vs. stubs vs. actuals as dependencies in unit tests.
- What is the current state of codebase?

Coders will do good to study:
- SOLID-principles and clean code.
- Team’s common agreed vocabulary for unit testing jargon.

Disclaimer:
- Our generics/TypeScript is a bit rusty.
- By decision, the unit testing as done here is *local* instead of *behavioral*. This is not something we can do forever, as coupling small unit tests with very specific technical details will in the long run hinder ability to refactor instead of enabling it. This decision was done to introduce first steps of a strategy, that can later grow to benefit from said behaviors.

Closing #3918

Commit:
Try stop linting before running tests

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Implement breadcrumbs for deleting different kube objects

This was done in a naive way by disregarding stuff such as:
- Avoidance of shared global state (dependency inversion principle)
- Unit testing
- Minimal implementation over defensive code (elvis, just in case)
- Full understanding for use cases of all different kube objects

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Rename method to not be misleading

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

jansav committed:
Demonstrate naive way to make component not responsible of it's dependencies and not use global shared state

This is to make a step towards satisfaction of Dependency Inversion Principle. It's not fully satisfied because we still depend on concrete instances of classes instead of interfaces.

This will start chain of thought to pursue warm and friendly unit testing and introduction of SOLID principles.

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

Commit:
Demonstrate a step further to have an interface as dependency instead of concrete instance

With this we get yet closer to satisfaction of Dependency Inversion Principle.

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Demonstrate even step further to have an interface as dependency instead of concrete instance

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Simplify a way of getting cluster name

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Make another dependency a segregated interface instead of concrete instance

Notice how interface is in fact segregated, i.e. it only advertises the minimal properties currently required.

With this we satisfy Interface Segregation Principle and thus permit e.g. minimal requirement for stubbing in future unit tests.

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Improve naming of type parameter

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Make typing of dependency functions more strict

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Introduce first unit test made possible because of recently demonstrated changes

Key changes were:
- Introduction of dependencies as interfaces to permit stubbing in general (Dependency Inversion Principle)
- Introduction of segregated interfaces to only require stubbing of minimal stuff (Interface Segregation Principle)

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Kludge first unit test with more than minimal data

This is a kludge because extraction of segregated interface for KubeObject is very difficult. This commit uses actual instance of KubeObject instead as a fake.

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Fix code style

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Iku-turso committed:
Implement first meaningful unit tests for a complete behavior

Note: in this case, the behavior is even asynchronous, and thus a tool "asyncFn" is used to keep the test as simple as possible.

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@gmail.com>

Commit:
Add todo for unit testing a weird complication without full understanding

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Bundle dependencies of a component to a composing object

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Simplify dependencies for a component

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Revert "Try stop linting before running tests"

This reverts commit d1fca3c38c4f47625053f4b8a307dca0ec7db528.

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Tweak UI in a dialog

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Update yarn.lock

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Add unit test for missing namespace when removing a kube object

We almost forgot to test this, because we started without TDD.

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@houston-inc.com>

Commit:
Switch to non deprecated prop type

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Remove unused data-attribute

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Fix code style

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Increase credibility of some unit tests by using actual instance as stub

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Remove prop from wrong place

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

jakolehm commented on .idea/.gitignore:
```diff
@@ -0,0 +1,6 @@
+# Default ignored files
```
Please split all `.idea/` files into separate PR.

jansav replied:
#4293

jakolehm commented on extensions/kube-object-event-status/package-lock.json:
```diff
@@ -16,9 +16,9 @@
-          "version": "7.14.0",
-          "resolved": "https://registry.npmjs.org/@babel/runtime/-/runtime-7.14.0.tgz",
-          "integrity": "sha512-JELkvo/DlpNdJ7dlyw/eY7E0suy5i5GQH+Vlxaq1nsNJ+H7f4Vtv3jMeCEgRhZZQFXTjldYfQgv2qmM6M1v5wA==",
+          "version": "7.16.0",
```
Idk why package-lock files within extensions were changed but these changes should not be part of this PR.

jansav replied:
True, removed.

jakolehm commented on src/renderer/components/kube-object-menu/inject-naive/inject-naive.tsx:
```diff
@@ -0,0 +1,54 @@
+}
+
+export const InjectNaive = ({ Component, ...props }: Props) => {
+  const kubeObjectMenuRegistry = KubeObjectMenuRegistry.getInstance();
```
Not sure if I understand why we need `InjectNaive`, couldn't `KubeObjectMenu` have default props which could be overridden?

jansav replied:
`InjectNaive` is there to represent the abstraction that should be created between components and it's dependencies. We have seen that usage of default props will limit our choices in the future. Good example would be that dependency is asynchronous.

Of course, in this first step, the `InjectNaive` is just hard-coded solution to get rid of shared global state in UI component itself. In future, we need find a way to inject those dependencies from somewhere else.

Nokel81 replied:
How could a dependency be asynchronous? React has a synchronous rendering pipeline.

jansav replied:
Relatively easily. Of course, in React prop should always be synchronous, but with little help of MobX, we can observe the status of a promise and render based on that. You can only imagine the possibilities that it opens. 

However, since it's only one of the benefits, I think it's too deep discussion to have here :) Quick huddle would be great so that we can discuss it further. 

Our experience says, that creating abstraction here opens a lot of improvements that will make the component itself lot simpler.

Iku-turso replied:
Indeed, historically asynchronous dependencies have been a wonderful tool to simplify some questionable code here and there. Also the separation of concerns implemented here rubs me the good way.

Let me partake in this huddle, ey guys? :)

Nokel81 replied:
A huddle might be a good idea. But thinking about this more I still don't see why leveraging default props isn't a better design or at least just as good. That way we don't introduce any new types

jansav replied:
One thing that comes my mind immediately is that when we get unit testing infrastructure to correct level, we need to start writing unit tests to larger scope to support refactoring in scale. 

That being said, it would cause the issue that you need to provide huge amount of dependencies as props from top level to component levels. 

But yeah, let's schedule huddle and go through it asap.

Nokel81 replied:
That is why I think we should also leverage default props, so that the users of a component don't need to change

jansav replied:
How would you handle following situation?

**You want to create unit test for a component which uses other components that have dependencies that need to be stubbed.**

For example:

- Render page in unit test
- One of the component in page has dependency that gets data from an API call
- You want to be able to control the result of said API call and see what effect it has on the page

Disclaimer: 
- API call is just an example and it's stupid one but could be anything that need to be stubbed in test
- jest.doMock works, but performance is horrible and it will cause issues when you have large test suite.
- Shared global state is bad and shouldn't be used here

To solve this, you would need to pass the dependency from unit test to top level (page) component to child component, but if there is an abstraction for injecting said dependency, you can just stub the dependency using the abstraction and all places that inject the dependency will have exactly same instance of it. This makes our unit test suite much more believable and robust.

Nokel81 replied:
I don't know if I buy the robustness argument since the effect is the same, and is believability really a benefit if you are just hiding where you inject the code?

jansav replied:
Not hiding of course. As said, it was not ready when we opened the draft for discussion. We are preparing one example how it could be done :)

chenhunghan replied:
One option to "inject dependencies" is to use custom renderer https://testing-library.com/docs/react-testing-library/setup/#custom-render

There are some real world examples in lenscloud/frontend/__test__. (Maybe not the best one...).

I think it's better to stop using defaultProps, React team seems to plan to retire it for functional components at some point in a major release.

jansav replied:
Check the proposed solution in #4338.

Commit:
Try altering cancelIdleCallback stub to see if that makes test deterministic

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

Commit:
Revert "Try altering cancelIdleCallback stub to see if that makes test deterministic"

This reverts commit d5148b6537944525414e918709d862d0593107d0.

Signed-off-by: Janne Savolainen <janne.savolainen@houston-inc.com>

chenhunghan commented on .gitignore:
```diff
@@ -17,3 +17,5 @@ types/extension-renderer-api.d.ts
 extensions/*/dist
 docs/extensions/api
 site/
+
```
[nit] Do we need this empty line?

chenhunghan commented on src/renderer/components/kube-object-menu/kube-object-menu.test.tsx:
```diff
@@ -0,0 +1,295 @@
+
+      removeActionMock = asyncFn();
+
+      ({ baseElement, getByTestId, queryByTestId } = render(
```
I think nowadays they recommend to use `screen` for queries (https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen) e.g. `screen.getByTestId`, so you probably don't need to have `let getByTestId` but better use `screen.getByTestId()` in each block.

jansav replied:
Good catch. Resolved

chenhunghan commented on src/renderer/components/kube-object-menu/kube-object-menu.test.tsx:
```diff
@@ -0,0 +1,295 @@
+    });
+
+    it("does not open a confirmation dialog yet", () => {
+      expect(queryByTestId("confirmation-dialog")).toBeNull();
```
See above comments.
```suggestion
      expect(screen.queryByTestId("confirmation-dialog")).toBeNull();
```

chenhunghan commented on src/renderer/components/kube-object-menu/kube-object-menu.test.tsx:
```diff
@@ -0,0 +1,295 @@
+
+  describe("given kube object", () => {
+    let baseElement: Element;
+    let removeActionMock: any;
```
[nit] Better typing
```suggestion
    let removeActionMock: ReturnType<typeof asyncFn>;
```

jansav replied:
Nice catch. :)

chenhunghan commented on src/renderer/components/kube-object-menu/kube-object-menu.test.tsx:
```diff
@@ -0,0 +1,295 @@
+  beforeEach(() => {
+    // TODO: Remove illegal global overwrites for what should be a dependency somewhere.
+    // TODO: Remove usage of experimental browser API.
+    window.requestIdleCallback = (callback: IdleRequestCallback): number => {
```
Do we need to reset this in `afterAll` or `afterEach`?

jansav replied:
Good question. In perfect world we shouldn't need to do that but not sure at the moment. So this is something that we need to verify.

nevalla commented on src/renderer/initializers/kube-object-detail-registry.tsx:
```diff
@@ -54,7 +54,7 @@ import { ReplicaSetDetails } from "../components/+workloads-replicasets";
 import type { KubeObjectDetailsProps } from "../components/kube-object-details";
 
-export function intiKubeObjectDetailRegistry() {
+export function initKubeObjectDetailRegistry() {
```
Good catch 👍  Although this could/should be in separate PR, since it's out of scope of this PR and probably would land to `master` faster.

jansav replied:
Well, true. I don't think that it's needed here. :) It will get there eventually.

Iku-turso replied:
Hmm. It could just be better to keep the overhead for trivial cleanup / boy scouting at minimum. It won't strain the PR if a reviewer simply skips/skims commits that are clearly trivial. And when it comes to clean-up, I would choose quantity over fast-tracking to `master`.

jansav committed:
Switch to using testing library queries from screen export

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Improving typing of mock

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Nokel81 commented on .gitignore:
```diff
@@ -17,3 +17,4 @@ types/extension-renderer-api.d.ts
 extensions/*/dist
 docs/extensions/api
 site/
+.prettierrc
```
We have been using `eslint` as our linter and formatter in this repo, so it seems strange to add prettier

jansav replied:
That's the reason that it's listed in `.gitignore`.  Prettier does lot more than `eslint` and I'm trying to find a way to use it to format code without messing up the existing code so that those who want to utilize it, can do so.

Iku-turso replied:
I agree with @jansav, current `eslint` configuration is a tad suboptimal, and causes manual work to keep code reasonably formatted. If optional usage of `prettier` even complies to `eslint`, then why not go for it?

Nokel81 commented on src/renderer/components/kube-object-menu/inject-naive/inject-naive.tsx:
```diff
@@ -0,0 +1,46 @@
+
+interface Props {
+  Component: React.ReactType<any>
+  [key: string]: any,
```
Also I really don't line this lack of type safety

jansav replied:
Well yeah, of course. But as said, `InjectNaive` is just hard-coded solution which is not ready to be merged. :) It's here just to pass the dependencies to `KubeObjectMenu` and therefore not scalable.

Iku-turso replied:
Type safety is now in place: #4338.

Nokel81 commented on src/common/k8s-api/api-manager.ts:
```diff
@@ -27,7 +27,11 @@ import type { KubeApi } from "./kube-api";
 import { IKubeObjectRef, parseKubeApi, createKubeApiURL } from "./kube-api-parse";
 
-export class ApiManager {
+export interface IGettableStore {
```
Personally I really dislike the `I` prefixing of interfaces. 

Is it also possible to have better names for these interfaces, currently they seem too generic.

Iku-turso replied:
On `I`-prefixing: Together with `T`-prefixing, this has many times proven valuable tool in making some more difficult typings understandable. Also it would be good to have the code somehow look off, when a coder is supposed to use an interface instead of concretion. With this, we have a better chance of making *Inversion of Control* and *loose coupling* happen on team level :)

On interface naming: Agreed. New interfaces added to better communicate and document intent.

Nokel81 replied:
I disagree, would you please provide some source for your claim? I find them distracting. Items should be obvious what they are from how they are used.

Iku-turso replied:
Which claim? :) A general goal here is to make stuff more obvious. Lack of Inversion of Control makes code less obvious by being a tangle of concrete dependencies. I would say that anything that could help in fixing this is a welcome thing to be tried.

Nokel81 replied:
This claim:

> On I-prefixing: Together with T-prefixing, this has many times proven valuable tool in making some more difficult typings understandable.

panuhorsmalahti replied:
I'd go with the commonly used convention in TypeScript projects, e.g. not using the `I`-prefix.

jweak replied:
I tend to agree. At least have only one way throughout the codebase. Either prefix all or nothing.

Iku-turso replied:
@panuhorsmalahti let's not make a vote out this, as best rationale should win. What is yours? Convention for sake of convention does not count :)

Nokel81 replied:
I have found that not having such prefixes limits the space of possible names so that name collisions happens more often within badly designed code.

panuhorsmalahti replied:
For code style questions commonly used conventions do matter, but are not the only factor. Following style that is used in, say, 90% of projects, makes it easier for new developers to jump in. Using something else requires extra arguments (that have been provided here). In any case, my own opinion is that the `I`-prefix makes the code less aesthetic and makes the code slightly more awkward to read and type. Prefixes do have a (minor) cost, otherwise we could add them to every single thing freely.

Iku-turso replied:
@Nokel81 
1. For example, we use T and I as key pieces of information when talking difficult subjects with other coders.

2. Also there was my other point:
> Also it would be good to have the code somehow look off, when a coder is supposed to use an interface instead of concretion. With this, we have a better chance of making Inversion of Control and loose coupling happen on team level :)

3. Here's a particularly dodgy piece of typing which debatably benefits from prefixes more than not:
```javascript
export interface IDependencyInjectionContainer {
    inject: <
      TInjectable extends IInjectable<TInstance, TDependencies>,
      TInstance,
      TDependencies,
    >(
      injectableKey: TInjectable,
    ) => ReturnType<TInjectable["instantiate"]>;
  }
```

jansav replied:
I think that we have common goal here: be able to separate interfaces from actual implementations. 

That being said, for me, it doesn't matter what is the solution to achieve that, but I haven't heard any better alternatives than prefixing them with something. Interfaces and implementations are separate for a reason, so how do you think we should separate them? Currently it's not so clear.

Nokel81 replied:
We should separate them by name, but we don't need to "namespace" the interfaces with prefixes.

jansav replied:
@Nokel81 Could you provide example? :)

Iku-turso replied:
@Nokel81 but what about the 3 arguments? :)

Iku-turso replied:
@panuhorsmalahti I agree, yet still find the other arguments more pressing.

After all, we are trying to make a big thing happen here, with Inversion of Control and all. Expect more eggs to be broken :)

Nokel81 replied:
1. When taking it is quite easy to say "an interface" or "a class/implementation" called "Foo", when working in an IDE, there is hover intellisense and "goto type definition" and such.
2. I would need to an example as to why prefixing helps. The general rule of "consume interfaces, provide implementations" doesn't need any naming conventions to be realized. 
3. Because of how typescript type parameters work all those prefixes are unnecessary, `T extends F` means that `F` must be some sort of interface, and the fact that `T` is used as a parameter type means that it must be an implementation.

Iku-turso replied:
@jweak we covered your point IRL. Nice to meet you! :)

Iku-turso replied:
@Nokel81

> When taking it is quite easy to say "an interface" or "a class/implementation" called "Foo"

That's too much overhead to be practical.

> when working in an IDE, there is hover intellisense and "goto type definition" and such.

One shouldn't be required to hover stuff to make code easier to read in this way.

> The general rule of "consume interfaces, provide implementations" doesn't need any naming conventions to be realized.

The good rule shouldn't need the conventions, but it sure does benefit from them, as transgressions get easier to spot.

> Because of how typescript type parameters work all those prefixes are unnecessary, `T extends F` means that `F` must be some sort of interface, and the fact that `T` is used as a parameter type means that it must be an implementation.

This comes with additional cognitive load, making it a losing bargain.

Nokel81 replied:
> That's too much overhead to be practical.

I disagree, saying the prefix is more cumbersome as it won't ever flow with the rest of the name

> One shouldn't be required to hover stuff to make code easier to read in this way.

Why not, having too much information at a time can make code hard to read so having an optional way of getting more information when necessary is very useful.

> The good rule shouldn't need the conventions, but it sure does benefit from them, as transgressions get easier to spot.

Maybe but that doesn't rule out the fact that this convention in particular (adding predefined prefixes and/or suffixes) is an eyesore. Beautiful code should be readable and understandable without this crutch.

This convention naturally leads to doing what C# and Java do with `FooImpl implements IFoo` which just is terrible.

> This comes with additional cognitive load, making it a losing bargain.

How is more cognitive load then the convention? The code reader still has to understand what is going on

jakolehm replied:
Most of the existing interfaces are without `I` prefix and we cannot really change those without breaking extension api types. So my recommendation is to use that convention for now. IMHO main thing that matters here is to have a convention.

Iku-turso replied:
@jakolehm That is a pressing argument! I might know some work-arounds, but I don't have enough visibility at this time.

Complying to convention and moving on! :)

Nokel81 commented on src/common/k8s-api/api-manager.ts:
```diff
@@ -27,7 +27,11 @@ import type { KubeApi } from "./kube-api";
 
-export class ApiManager {
+export interface IGettableStore {
+  getStore<TKubeObjectStore extends KubeObjectStore<KubeObject>>(api: string | KubeApi<KubeObject>): TKubeObjectStore | undefined;
```
@jakolehm I recently stumbled across the following rule https://github.com/microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md

Would it be a good idea to slowly move towards that?

Iku-turso replied:
Makes sense to me :)

jansav committed:
Demonstrate how InjectNaive could be not naive by using injectables

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Fix incorrect typing in test

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Update snapshots

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Organize some interfaces to communicate intent

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Try making UI unit testable in Ubuntu by removing usage of component which uses an experimental browser feature

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Update version of injectable for disambiguous typing

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Iku-turso committed:
Make unit test runs faster and not fail for TypeScript errors/warnings

Signed-off-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Fix merge conflicts

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Make component transient

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Extract menu items as own dependency

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Add types for ogre-tools in WebStorm

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Try extracting dependencies so separate files

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Simplify typing of async mock

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Iku-turso committed:
Update ogre-tools for a more subtle way of providing components with dependencies

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>
Signed-off-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Switch to using "container" instead of Inject-component for following reasons:

- Conservation of Extension API, as now injection mechanism travels where ever component is used with unchanged interface in place
- Minimize impact on existing code by replacing duplicated Inject-components with usages of single container

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Revert "Try extracting dependencies to separate files"

This reverts commit 8af0e748

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Rename files for convention

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Remove changes to file that didn't need to change

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Remove test that is not needed anymore

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Remove todo

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

Nokel81 commented on src/extensions/registries/kube-object-menu-registry.ts:
```diff
@@ -32,10 +32,10 @@ export interface KubeObjectMenuRegistration {
 }
 
+
 export class KubeObjectMenuRegistry extends BaseRegistry<KubeObjectMenuRegistration> {
```
I don't think this change is required for this change.

jansav replied:
Reverted changes in this file.

Nokel81 commented on src/renderer/components/kube-object-menu/index.ts:
```diff
@@ -19,4 +19,5 @@
 
-export * from "./kube-object-menu";
+export type { KubeObjectMenuProps } from "./kube-object-menu";
+export * from "./kube-object-menu-container";
```
I think we might as well do `export { KubeObjectMenu } ...`

jansav replied:
Done

Nokel81 commented on src/renderer/components/kube-object-menu/kube-object-menu.tsx:
```diff
@@ -53,61 +65,61 @@ export class KubeObjectMenu<T extends KubeObject> extends React.Component<KubeOb
 
   @boundMethod
-  renderRemoveMessage() {
+  getRemoveMessage() {
```
This is technically a breaking change

jansav replied:
Hmm, technically maybe. 

Someone might extend the component and try accessing old method (or override it). 
Can you do it? Yes. Should you do it? No. 

Our ability to refactor anything is kinda limited, if we decide that we want to keep supporting that kind of use case. If so, I think that we should mark this component as `@deprecated` and create completely new version of `KubeObjectMenu`.

What's your opinion?

jansav replied:
Side note: 

We could use some sort of mechanism to guard the Extension API from sudden changes such as this one. Currently there needs to be too much manual care and convention to be sure about conservation of the API. 

How about flurry of interfaces that communicate stake in Extension API? That way we would know which parts of code base are open for healthy refactoring.

Nokel81 replied:
I completely agree that it is currently hard to tell what the extension API both "is" and what is allowed to change.

> Our ability to refactor anything is kinda limited, if we decide that we want to keep supporting that kind of use case. If so, I think that we should mark this component as @deprecated and create completely new version of KubeObjectMenu.
> 
> What's your opinion?

I don't think we need to for this PR, looking at the change I don't actually see why you want to change the method name either. Our standard practice is to name method `render*` if they return a `React.ReactNode` which is what this one does.

jansav replied:
The naming might be sub-optimal imho, but breaking Extension API is unacceptable. Reverted.

Nokel81 commented on src/renderer/components/kube-object-menu/kube-object-menu.tsx:
```diff
@@ -53,61 +65,61 @@ export class KubeObjectMenu<T extends KubeObject> extends React.Component<KubeOb
+      <MenuItem
+        object={object}
+        toolbar={toolbar}
+        // TODO: Fix misuse of index in key
```
Isn't this only a misuse if the items will be reordered?

jansav replied:
I would avoid using index if UID is available as a convention but your argument is valid, so I removed the comment.

Nokel81 commented on src/renderer/components/menu/menu-actions.tsx:
```diff
@@ -125,32 +124,31 @@ export class MenuActions extends React.Component<MenuActionsProps> {
     return (
       <>
         {this.renderTriggerIcon()}
-        <RenderDelay>
```
@aleksfront Wasn't this rendering delay added on purposes to help with initial load times?

jansav replied:
We discussed with @aleksfront about the requirement of `RenderDelay` and came to decision that it's not needed. We didn't see any impact on initial load times.

jansav committed:
Remove even more changes that were not needed after all

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Make export explicit

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Remove todo

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Revert breaking change in Extension API

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

Nokel81 commented on src/renderer/components/kube-object-menu/kube-object-menu.test.tsx:
```diff
@@ -0,0 +1,299 @@
+import { KubeObjectMenu } from "./index";
+
+describe("kube-object-menu", () => {
+  let objectStub: KubeObject | null;
```
nit, I don't think this needs to be here. All the uses are contained within each function call

jansav replied:
Good catch, simplified.

jansav committed:
Simplify test

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

jansav committed:
Remove changes to file that didn't need to change

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Update asyncFn

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
